src/librustc_middle/ty/query/keys.rs

```diff
@@ -296,3 +296,14 @@ impl Key for (Symbol, u32, u32) {
         DUMMY_SP
     }
 }
+
+impl<'tcx> Key for (ty::ParamEnv<'tcx>, DefId, SubstsRef<'tcx>) {
```

Comment:
I guess a strong hint is that you had to add a new `Key` impl: generally, you shouldn't need one.